JSON output for fix reaxff/species delete keyword

jrgissing opened the pull request:
**Summary**

New ‘fix reaxff/species' feature that produces JSON output for ReaxFF that includes atom IDs, types and coords of deleted molecules. ‘JSON mode’ is triggered by detecting the extension of the name of delete species output file (e.g., ‘deleted_molecule.json’).

Thanks to Jacob Breese (OSU) for testing

**Related Issue(s)**

template for general JSON output from run command, so some discussion necessary for format. see attached for example file of deleted water molecules

might be wise to add to the general LAMMPS output howto about JSON per-timestep output, or once it is added for a few commands

related to #4625

**Author(s)**

Jake Gissinger (AFRL, Stevens)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

[species-del.json](https://github.com/user-attachments/files/23152402/species-del.json)

[example_deleting_water.zip](https://github.com/user-attachments/files/20958879/example_deleting_water.zip)

jrgissing committed:
framework for JSON delete keyword

needs actual output of molecules

jrgissing committed:
use utils for json extension  matching

might be better to still have 'has_extension' wrapper in platform, due to ease of and grepping for it

jrgissing committed:
working v0

jrgissing committed:
better version

jrgissing committed:
lost a bracket somewhere along the way

akohlmey commented on src/output.cpp:
```diff
@@ -628,6 +630,146 @@ void Output::write_restart(bigint ntimestep)
+  unique_ivec.erase(0);
+  std::vector<int> mivec(unique_ivec.begin(), unique_ivec.end());
+
+  int approxsize = atom->natoms / comm->nprocs;
```
Why not use `atom->nmax`?

akohlmey commented on src/label_map.h:
```diff
@@ -43,10 +43,11 @@ class LabelMap : protected Pointers {
   void read_restart(FILE *fp);
   void write_restart(FILE *);
 
- protected:
```
Rather than making data members public (as was done in the past), we should follow modern C++ guidelines for new code and add accessor functions.

akohlmey commented on src/output.cpp:
```diff
@@ -628,6 +630,146 @@ void Output::write_restart(bigint ntimestep)
+    offsets[2] = offsetof(Particle, x);
+    MPI_Datatype types[nfields] = {MPI_INT, MPI_INT, MPI_DOUBLE};
+
+    MPI_Type_create_struct(nfields, blocklengths, offsets, types, &ParticleStructType);
```
This needs to be implemented in the STUBS library for serial compilation.

jrgissing committed:
address Axel's comments

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
tweaks

jrgissing committed:
Merge branch 'fix-reaxff-speces_JSON-output' of https://github.com/jrgissing/lammps into fix-reaxff-speces_JSON-output

jrgissing committed:
whitespace

jrgissing committed:
docs

jrgissing committed:
typos

jrgissing committed:
undo mysterious line deletion

jrgissing committed:
need to turn auto whitespace cleanup back on

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
uniquify property/atom property name

fix for change from circa Jan 2025 that broke backward compatibility by preventing multiple "fix reaxff/species" from being defined

jrgissing commented:
I propose to change my current suggested "deleted_molecules" key to simply "molecules", perhaps adding a description subkey to identify them as deleted. The idea is for it to be the same all fixes that will output molecules, to make parsing more universal. in particular, we would like to visualize this output, and visualization programs cannot be expected to parse different keys for each fix

jrgissing committed:
simplify JSON key

jrgissing commented:
note: I changed the "deleted_molecules" key to simply "molecules". the "subformat" key identifies the molecules as deleted.

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
Merge branch 'lammps:develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
fix memory leak

jrgissing committed:
Merge branch 'lammps:develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

akohlmey commented on doc/src/fix_reaxff_species.rst:
```diff
@@ -140,28 +140,88 @@ be either a list of specific chemical formulae or a range of molecular
+
+   {
+       "application": "LAMMPS",
+       "format": "output",
```
@jrgissing to make this more consistent with the rest of LAMMPS and also be more future-proof, I suggest we call this format "dump", and the "subformat" (or perhaps "style"?) should be called something like "by-molecule" or "by-atom" to indicate how the data is grouped. The "fix reaxff/species: delete keyword" string would be more fitting for a "title". Then I would rather have "timesteps" instead of "run_output", so that "run_output": [ {"timestep": 1, "molecules" [...]},...] could become "timesteps": [{10, [...]}, 20, [...]},...].

The "by-atom" format would then be more like the conventional dump styles, only that rather than having one big blob per timestep, we would have sections like "type", "coord", "charge", "molecule", "force", "vel", "mass", ... 
There is already a by column version with dump style yaml.

akohlmey replied:
The by-atom format could become something of an alternative to a data file, if topology data and type-label mapping would be included.

jrgissing replied:
I agree with all these suggestions. For molecule vs atom, what about:
"grouping": "molecule"

akohlmey replied:
hmmm... in the "by-atom" format, the "grouping" of information would technically speaking be by property.  So in that sense a keyword like "order" or "ordering" would be better. But to be perfectly frank, I am not 100% comfortable with either of the choices that you or I have mentioned (including "sub-format").

Now that I look at this for a second time, I am beginning to like something like "style": "atom" and "style": "molecule" more. *Definitely* without the "by-"; that was a bad idea.  It is cleaner to just have a single word for a keyword. I suggest you pick (for now) what *you* like best for as long as it is understood that it is not yet the final word and I can try, if I can come up with something I like 100% (and can convince you of the same).

Another idea (primarily for a future dump style json) would be to have a flag requesting to output only fields that are *different* from the previous step. That could be a big space saver for simulations with partially frozen content or when outputting topology data (like changing bonds/angles/dihedrals) for some atoms. That kind of format would need another keyword to indicate whether unchanged data is skipped or not. 

We would need to write down the final format(s) in a `tools/json/dump-schema.json` file. That can be done in multiple iterations if needed. we should just converge for the next feature release (planned for mid-to-late December if development remains as slow as the last weeks and I am spending as much time on writing papers as I did since the last release).  When I was adding JSON support to the molecule command, I was writing both at the same time and used the JSON schema to make sure my C++ code was doing what was intended and used the JSON schema as a form of documentation of the format specifics before writing the actual docs.

jrgissing replied:
what about:
  "format": {
    "type": "dump",
    "output": "molecules"
  }

akohlmey replied:
No, we should make the global keywords have the same layout for the different formats.
So it would be better to have:
```json
"format": "dump",
"output": "molecules",
```
and:
```json
"format": "dump",
"output": "atoms",
```

or use "style" instead of "output"... ;-)

jrgissing replied:
OK, made those updates. Note that I would vote for leaving the 'timestep' key inside the 'timesteps' object for human readability and searchability. I should be able to get to the schema this weekend.

jrgissing committed:
Merge branch 'lammps:develop' into fix-reaxff-speces_JSON-output

jrgissing committed:
json key updates

jrgissing committed:
json key updates

jrgissing committed:
docs+tools update

jrgissing commented:
@akohlmey ok, this is ready for another look

jrgissing committed:
Merge branch 'develop' into fix-reaxff-speces_JSON-output

jtclemm commented on doc/src/fix_reaxff_species.rst:
```diff
@@ -140,28 +140,90 @@ be either a list of specific chemical formulae or a range of molecular
+Nlimit within an interval of Nsteps timesteps. Nlimit can be specified with
+an equal-style :doc:`variable <variable>`. When using the
+*delete_rate_limit* keyword, no deletions are permitted to occur within the
+first Nsteps timesteps of the first run (after reading a either a data or
```
Extra word? "after reading *a* either a data or"

jtclemm commented on doc/src/fix_reaxff_species.rst:
```diff
@@ -140,28 +140,90 @@ be either a list of specific chemical formulae or a range of molecular
+The *delete* keyword can output information about the deleted molecules in
+either legacy format or JSON format. The latter is activated when the
+*filedel* argument has a '.json' extension.  The legacy format lists how
+many of each species is deleted, while the JSON format provides the atom ID,
```
Are deleted, I think

jtclemm reviewed (approved):
I'm not the best reviewer since I'm not familiar with JSON nor am I super familiar with the utilities in LAMMPS for file writing, but everything else looked good to me. Just spotted a few minor proofreading edits.

jrgissing committed:
Update fix_reaxff_species.rst

akohlmey commented on tools/json/dump-molecules-schema.json:
```diff
@@ -0,0 +1,121 @@
+            "type": "string",
+            "const": "dump"
+        },
+        "style": {
```
Thanks for providing the JSON schema file. I think we should change the "style" entry to
```JSON
"style": {
    "enum": ["molecules"]
},
```
That would make it easier to extend this for other JSON dump styles and we would only need one schema file.

jrgissing replied:
Are you suspecting that additional keys within 'style' will be necessary? Do you have any examples in mind? I am already suspecting that certain fix dumps will need to include fix-specific metadata on a per-molecule basis.

akohlmey replied:
I would prefer to have one single schema describing a general JSON dump format. As mentioned in an earlier discussion, I can imagine that we also implement a dump style json and that could have the data organized similar to dump style custom or dump style yaml in columns and listed by atom ID (aka "style": "atom") or it could be ordered by property (-> "style": "property"), i.e. each section is organized more like a molecule JSON file or a data file with "coord", "type", "molecule", "charge", "velocity", "force", etc. sections and each containing data for all atoms.

akohlmey commented on tools/json/dump-molecules-schema.json:
```diff
@@ -0,0 +1,121 @@
+                "type": "object",
+                "required": ["timestep", "molecules"],
+                "properties": {
+                    "timestep": {
```
I understand your desire for readability, but I have concerns about how we can efficiently access the parsed data. If the timestep data is just a list of integer + object, we can easily (and quickly) access a timestep by its value and extract the list of all included timesteps. By making it a "timestep": value / "molecules": list, you would have to search through *all* steps to find a specific one, if I am not mistaken. We could require starting at step 0 and equal distancing, but that would be a limitation, that was not so long ago removed from (most of) the regular dump styles.

jrgissing replied:
ok

akohlmey commented on tools/json/dump-molecules-schema.json:
```diff
@@ -0,0 +1,121 @@
+                      "type": "integer",
+                      "minimum": 0
+                    },
+                    "molecules": {
```
It may not be needed and have much meaning for ReaxFF output, but should the molecules have molecule IDs (for example when outputting a trajectory by molecule from a future dump style json)?

jrgissing replied:
This may be useful for future fix-bond dumps, but ReaxFF uses atom_style charge, so I don't you can assign molecule IDs in the first place, nor hope to keep consistent values during reactions. Seems like a v2 addition

akohlmey replied:
fix reaxff/species computes the clusterID array which could be used in place of a molecule ID. Thuse will change, when there are reactions, of course, but without they are consistent, since the order of the clusterIDs is derived from the smallest atom-ID in the cluster. We have at times recommended people to overwrite the molecule IDs with that information.

akohlmey replied:
At any rate, the molecule ID would be an optional field and thus can be omitted, if so desired or for situations where the information is not available.

akohlmey commented on tools/json/dump-molecules-schema.json:
```diff
@@ -0,0 +1,121 @@
+                                    }
+
+                                },
+                                "bonds": {
```
This probably should also include angles, dihedrals, and impropers.
And how about other properties supported by molecule files?
Or Should we wait until we have a proper use case and then declare a revision 2 of the format?

jrgissing replied:
I think angles, dihedrals, impropers could be a possible v2 thing. I added bonds because fixed-bond applications would certainly need to include those.

jrgissing replied:
or, perhaps a better option is to simply use a $ref to our LAMMPS molecule JSON schema, to clearly support the full (and potentially changing) molecule spec. We should just make sure that per-molecule metadata could be associated with each molecule.

akohlmey replied:
Interesting idea. In that case there could be a Molecule class instance that is cleared and filled with available data for each molecule, then you can call `Molecule::to_json()` and with the thus created json class instance output the resulting serialized data structure using `json::dump(2)` and only create the JSON output manually for the outer loops. That would automatically apply the molecule JSON schema, but would have the downside of lots of redundant output.

jrgissing replied:
I was only referring to modifying the schema. So far, I like the idea of having full control of the output. Either way, I propose to modify the 'molecules' object to put 'LAMMPS molecule objects' inside a data key:
``` 
           "molecules": [
                {
                    "data": {

                        "types": {
                            ...
                        },
                        "coords": {
                            ...
                        }
                    }
                },
```

jrgissing committed:
Update dump-molecules-schema.json

JacksonBurns commented:
Hi @jrgissing this is a follow up to your talk at AIChE 2025, where you requested audience input about LAMMPS/JSON integration.

I'm far from an expert user, but I think it might be useful to offer some compatibility with rdkit's Molecule JSON format, [the docs for which are here](https://www.rdkit.org/docs/RDKit_Book.html#rdkitjson-format). Not sure if your current spec is compatible already, or if some converter would be needed, but just putting it out there. Great talk!

jrgissing commented:
> I'm far from an expert user, but I think it might be useful to offer some compatibility with rdkit's Molecule JSON format, [the docs for which are here](https://www.rdkit.org/docs/RDKit_Book.html#rdkitjson-format). Not sure if your current spec is compatible already, or if some converter would be needed, but just putting it out there. Great talk!

Thanks for the suggestion! Since it is not directly related to this PR, it would help to convert this suggestion into its own Issue
